cpp/src/lists/extract.cu

```diff
@@ -16,103 +16,131 @@
+ * the column returned is:                      `{5, -4, 3, -2, 1, MAX_SIZE_TYPE}`.
+ * All null indices are replaced with `MAX_SIZE_TYPE = numeric_limits<size_type>::max()`.
+ *
+ * The returned column can then be used to construct a lists column, for use
+ * with `segmented_gather()`.
  */
-template <bool PositiveIndex = true>
-struct map_index_fn {
-  column_device_view const d_offsets;  // offsets to each sublist (including validity mask)
-  size_type const index;               // index of element within each sublist
-  size_type const out_of_bounds;       // value to use to indicate out-of-bounds
-
-  __device__ int32_t operator()(size_type idx)
-  {
-    if (d_offsets.is_null(idx)) return out_of_bounds;
-    auto const offset = d_offsets.element<int32_t>(idx);
-    auto const length = d_offsets.element<int32_t>(idx + 1) - offset;
-    if (PositiveIndex)
-      return index < length ? index + offset : out_of_bounds;
-    else
-      return index >= -length ? length + index + offset : out_of_bounds;
-  }
-};
+std::unique_ptr<cudf::column> make_index_child(column_view const& indices,
+                                               size_type ignore,  // Placeholder for number of rows.
+                                               rmm::cuda_stream_view stream)
+{
+  // New column, near identical to `indices`, except with null values replaced.
+  // `segmented_gather()` on a null index should produce a null row.
+  if (not indices.nullable()) { return std::make_unique<column>(indices, stream); }
+
+  auto const indices_device_view = column_device_view::create(indices);
+  auto const d_indices           = *indices_device_view;
+  auto const null_index          = std::numeric_limits<size_type>::max();
+  auto const null_replaced_iter_begin =
+    cudf::detail::make_null_replacement_iterator(d_indices, null_index);
+  auto index_child = cudf::make_numeric_column(
+    data_type{type_id::INT32}, indices.size(), mask_state::UNALLOCATED, stream);
+  thrust::copy_n(rmm::exec_policy(stream),
+                 null_replaced_iter_begin,
+                 indices.size(),
+                 index_child->mutable_view().begin<size_type>());
+  return index_child;
+}
+
+/**
+ * @brief Helper to construct a column of indices, for use with `segmented_gather()`.
+ *
+ * When indices are specified as a size_type, e.g. `7`,
+ * the column returned is: `{ 7, 7, 7, 7, 7 }`.
+ *
+ * The returned column can then be used to construct a lists column, for use
+ * with `segmented_gather()`.
+ */
+std::unique_ptr<cudf::column> make_index_child(size_type index,
+                                               size_type num_rows,
+                                               rmm::cuda_stream_view stream)
+{
+  auto index_child =  // [index, index, index, ..., index]
+    make_numeric_column(data_type{type_id::INT32}, num_rows, mask_state::UNALLOCATED, stream);
+  thrust::fill_n(
+    rmm::exec_policy(stream), index_child->mutable_view().begin<size_type>(), num_rows, index);
```

Comment:
There is also a `cudf::detail:fill`
https://github.com/rapidsai/cudf/blob/1424a2dfce0f1015675cde82e9bef69533d1ada5/cpp/include/cudf/detail/fill.hpp#L45-L51

Sorry, I should've noticed that before.

Reply:
I think you can disregard this. I think the `thrust` one is fine here.